### ISSUE TITLE
Data Parallel without excessive gathering and scattering 

### ISSUE THREAD

**Issue by danieltudosiu** [Project-MONAI/MONAI#1143]
**Is your feature request related to a problem? Please describe.**
I am using quite big models that have very little room for loss functions, but at the same time I am using a loss function [1] that consumes more than your average L1/L2.

**Describe the solution you'd like**
There has been a quick fix by Hang Zhang [2] which basically stops the gathering of the network output and then has another DataParallel that contains the loss and continues the computation on the parallel GPUs. This avoids the issues scribbled here [3] where PyTorch uses one too many gather-scatter resulting in unbalanced GPUs usage.

**Describe alternatives you've considered**
An alternative is to break the paradigm that loss is not part of the network and cover the model and loss in a torch.nn.Module and feed that one in DataParallel.

**Additional context**
[1] https://github.com/jonbarron/robust_loss_pytorch
[2] https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/parallel.py
[3] https://discuss.pytorch.org/t/how-pytorchs-parallel-method-and-distributed-method-works/30349/2?u=danieltudosiu

**Comment by Nic-Ma** [comment]
Hi @danieltudosiu ,

Thanks for your interest here.
May I know what issue or question you are facing with MONAI? Or some suggestions for us?

Thanks.

**Comment by danieltudosiu** [comment]
Hi @Nic-Ma,

I think this is an edge case that would be good for Monai to cover as it will be encountered in medical imaging when we discuss:
 - Generative modeling
 - Natural language processing
 - 7T imaging tasks

Furthermore, this will further simply how DataParallel works and make it more intuitive instead of having a reduce step to a GPU before loss calculation. This, in turn, will benefit memory-constrained tasks to maximize the batch size as the GPU usage will become even across all GPUs.

May I know what issue or question you are facing with MONAI? <- Not with MONAI in particular, but I think this is an edge case that can be encountered in our domain.

 Or some suggestions for us? <- To integrate the solution from [2] and maintain it as part of MONAI. 

Thanks

**Comment by Nic-Ma** [comment]
Hi @danieltudosiu ,

OK, I see, thanks very much for your suggestion!
Mark as a discussion topic for next MONAI working group meetings.

Thanks.

**Comment by vikashg** [comment]
closing because of inactivity
